pubtools/pulplib/_impl/model/maintenance.py

```diff
@@ -0,0 +1,179 @@
+import logging
+import datetime
+import json
+import os
+
+import jsonschema
+
+from pubtools.pulplib._impl import compat_attr as attr
+from ..schema import load_schema
+from .common import InvalidDataException
+
+
+LOG = logging.getLogger("pubtools.pulplib")
+
+
+USER = os.environ.get("USER")
+HOSTNAME = os.environ.get("HOSTNAME")
+OWNER = "%s@%s" % (USER, HOSTNAME)
+
+
+def iso_time_now():
+    return datetime.datetime.utcnow().strftime("%Y-%m-%dT%H:%M:%SZ")
+
+
+@attr.s(kw_only=True, frozen=True)
+class MaintenanceEntry(object):
+    """Details about the maintenance"""
+
+    id = attr.ib(default=None, type=str)
+    """ID of repository in maintenance"""
+    message = attr.ib(default=None, type=str)
+    """Why this repository is in maintenance"""
+    owner = attr.ib(default=None, type=str)
+    """Who set this repository in maintenance mode"""
+    started = attr.ib(default=None, type=datetime.datetime)
+    """:class:`~datetime.datetime` in UTC at when the maintenance started"""
+
+
+@attr.s(kw_only=True, frozen=True)
+class MaintenanceReport(object):
+    """Represents the maintenance status of Pulp repositories
+
+    On release-engineering Pulp servers, it's possible to put individual repositories
+    into "maintenance mode".  When in maintenance mode, external publishes of a repository
+    will be blocked.  Other operations remain possible.
+
+    This object holds information on the set of repositories currently in maintenance mode.
+    """
+
+    _SCHEMA = load_schema("maintenance")
+
+    last_updated = attr.ib(default=iso_time_now(), type=datetime.datetime)
+    """:class:`~datetime.datetime` in UTC when this report was last updated,
+    if it's the first time the report is created, current time is used."""
+
+    last_updated_by = attr.ib(default=None, type=str)
+    """Person/party who updated the report last time"""
+
+    entries = attr.ib(default=attr.Factory(tuple), type=tuple)
+    """A tuple contains :class:`MaintenanceEntry` objects, indicates
+    which repositories are in maintenance mode and details.
+    If empty, then it means no repositories is in maintenance mode.
+    """
+
+    @classmethod
+    def from_data(cls, data):
```

Comment:
I would rather not make this one public API (or do you have a particular reason to do so?)

For the other PulpObject which do have a from_data, the main reason was because there's a lot of existing code for querying those objects already which aren't using this library, and it's valuable to be able to connect up with that code (e.g. use Repository class with existing code querying Pulp via pa_tool).

But for maintenance mode, I don't see much reason to provide a way for any other code to parse the raw data at all.  Allowing others to do that will essentially lock us in to the specific format of repos.json, as we couldn't change it without breaking this API.  I would rather avoid that, and try to have all of the code understanding that format built into this library only, as much as possible.

So, could we make this private too unless you have specific use-case for it?

(Note if I'm wrong about this, we can always promote it to public easily later.)